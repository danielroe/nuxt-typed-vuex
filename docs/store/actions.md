---
---

# Actions

Actions are async functions that have access to the Vuex instance and are passed a context object and optional payload.

This package provides a helper function to reduce boilerplate: `actionTree`. This function adds typings and returns the actions passed to it, without transforming them.

**Warning**: If you use the helper function, only the `commit` method is typed - and only for mutations within the module. Anything else should be accessed in a type-safe way through `this.app.$accessor`.

```ts
// Vanilla

export const actions = {
  async resetEmail(
    this: Store<RootState>,
    { state, commit }: ActionContext<RootState, RootState>,
    payload: string
  ) {
    // Typed
    state.email

    // Not typed - avoid
    commit('initialiseStore')

    // Typed
    this.app.$accessor.initialiseStore()
  },
}

// Helper function

import { actionTree } from 'nuxt-typed-vuex'

export const actions = actionTree(
  { state, getters, mutations },
  {
    async resetEmail({ commit, dispatch, getters, state }) {
      // Typed
      commit('initialiseStore')
      let a = getters.email
      let b = state._email

      // Not typed
      dispatch('resetEmail')

      // Typed
      this.app.$accessor.resetEmail()
    },
  }
)
```

**Note**

1. Even if you do not use the `actionTree` helper function, make sure not to use the `ActionTree` type provided by Vuex. This will interfere with type inference. You won't lose out by omitting it, as Typescript will complain if you pass an incompatible action into [the `getAccessorType` function](/setup.html#add-type-definitions).

2. This package does not support [object-style dispatches](https://vuex.vuejs.org/guide/actions.html).
