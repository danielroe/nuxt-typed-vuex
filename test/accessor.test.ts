import { useAccessor, getAccessorType, getAccessorFromStore } from '../'
import Vuex, { Store } from 'vuex'
import Vue from 'vue'
import { getters, state, actions, mutations } from './fixture/store'

import * as submodule from './fixture/store/submodule'

const pattern = {
  getters,
  state,
  actions,
  mutations,
  modules: {
    submodule: {
      ...submodule,
      namespaced: true,
      modules: {
        nestedSubmodule: {
          ...submodule,
          namespaced: true,
        },
      },
    },
  },
}

const accessorType = getAccessorType(pattern)
const submoduleAccessorType = getAccessorType(submodule)

const submoduleBehaviour = (accessor: typeof submoduleAccessorType) => {
  expect(accessor.firstName).toEqual('')
  accessor.setFirstName('Nina')
  expect(accessor.firstName).toEqual('Nina')
  accessor.setLastName('Willis')
  expect(accessor.fullName).toEqual('Nina Willis')

  accessor.initialise()
  expect(accessor.fullName).toEqual('John Baker')
  accessor.setName('Jordan Lawrence')
  expect(accessor.firstName).toEqual('Jordan')
}

describe.only('accessor', () => {
  let store: Store<any>
  let accessor: typeof accessorType

  beforeEach(() => {
    Vue.use(Vuex)
    store = new Store(pattern)
    accessor = useAccessor(store, pattern)
  })
  test('store exists', async () => {
    expect(pattern).toMatchSnapshot()
  })
  test('accessor object exists', async () => {
    expect(accessor).toMatchSnapshot()
  })
  test('accessor can be generated from store object', async () => {
    const accessorFromStore = getAccessorFromStore(store)(store)
    expect(accessorFromStore).toMatchSnapshot()
  })
  test('accessor works with state object', async () => {
    const stateAccessor = useAccessor(store, {
      state: {
        email: 'this does not matter',
      },
      mutations: {
        setEmail(state, newValue: string) {
          ;(state as any).email = newValue
        },
      },
    })
    expect(stateAccessor.email).toMatchSnapshot()
    stateAccessor.setEmail('test1@email.com')
    expect(stateAccessor.email).toMatchSnapshot()
  })
  test('accessor state, getter, mutation and actions work', async () => {
    expect(accessor.fullEmail).toEqual('')

    accessor.setEmail('john@j.com')
    expect(accessor.email).toEqual('john@j.com')
    expect(accessor.fullEmail).toEqual('john@j.com')

    accessor.resetEmail()
    expect(accessor.fullEmail).toEqual('a@a.com')
  })
  test('accessor submodule state, getter, mutation and actions work', async () => {
    submoduleBehaviour(accessor.submodule)
  })
  test('namespaced modules work', async () => {
    const nonNamespacedPattern = {
      getters,
      state,
      actions,
      mutations,
      modules: {
        submodule: {
          ...submodule,
          namespaced: false,
        },
      },
    }
    const nonNamespacedStore = new Store(nonNamespacedPattern)
    const nonNamespacedAccessor = useAccessor(
      nonNamespacedStore,
      nonNamespacedPattern
    )
    submoduleBehaviour(nonNamespacedAccessor.submodule)
  })
  test('nested modules work', async () => {
    submoduleBehaviour(accessor.submodule.nestedSubmodule)
  })
  test('namespaced dynamic modules work', async () => {
    store.registerModule('dynamicSubmodule', { ...submodule, namespaced: true })
    const dynamicAccessor = useAccessor(store, {
      modules: { dynamicSubmodule: { ...submodule, namespaced: true } },
    })

    submoduleBehaviour(dynamicAccessor.dynamicSubmodule)

    store.unregisterModule('dynamicSubmodule')
  })
  test('dynamic module helper function works', async () => {
    store.registerModule('dynamicSubmodule', submodule)
    const dynamicAccessor = useAccessor(store, submodule, 'dynamicSubmodule')

    submoduleBehaviour(dynamicAccessor)
    store.unregisterModule('dynamicSubmodule')
  })
})
