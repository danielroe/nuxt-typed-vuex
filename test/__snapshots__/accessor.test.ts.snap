// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessor accessor can be generated from store object 1`] = `
Object {
  "initialiseStore": [Function],
  "resetEmail": [Function],
  "resetEmailWithOptionalPayload": [Function],
  "setEmail": [Function],
  "submodule": Object {
    "initialise": [Function],
    "nestedSubmodule": Object {
      "initialise": [Function],
      "setFirstName": [Function],
      "setLastName": [Function],
      "setName": [Function],
    },
    "setFirstName": [Function],
    "setLastName": [Function],
    "setName": [Function],
  },
}
`;

exports[`accessor accessor object exists 1`] = `
Object {
  "initialiseStore": [Function],
  "resetEmail": [Function],
  "resetEmailWithOptionalPayload": [Function],
  "setEmail": [Function],
  "submodule": Object {
    "initialise": [Function],
    "nestedSubmodule": Object {
      "initialise": [Function],
      "setFirstName": [Function],
      "setLastName": [Function],
      "setName": [Function],
    },
    "setFirstName": [Function],
    "setLastName": [Function],
    "setName": [Function],
  },
}
`;

exports[`accessor accessor works with state object 1`] = `""`;

exports[`accessor accessor works with state object 2`] = `"test1@email.com"`;

exports[`accessor store exists 1`] = `
Object {
  "actions": Object {
    "resetEmail": [Function],
    "resetEmailWithOptionalPayload": [Function],
  },
  "getters": Object {
    "email": [Function],
    "fullEmail": [Function],
  },
  "modules": Object {
    "submodule": Object {
      "actions": Object {
        "initialise": [Function],
        "setName": [Function],
      },
      "getters": Object {
        "fullName": [Function],
      },
      "modules": Object {
        "nestedSubmodule": Object {
          "actions": Object {
            "initialise": [Function],
            "setName": [Function],
          },
          "getters": Object {
            "fullName": [Function],
          },
          "mutations": Object {
            "setFirstName": [Function],
            "setLastName": [Function],
          },
          "namespaced": true,
          "state": [Function],
        },
      },
      "mutations": Object {
        "setFirstName": [Function],
        "setLastName": [Function],
      },
      "namespaced": true,
      "state": [Function],
    },
  },
  "mutations": Object {
    "initialiseStore": [Function],
    "setEmail": [Function],
  },
  "state": [Function],
}
`;
